### ISSUE TITLE
-Wpointer-type-mismatch in arch/x86/include/asm/kvm-x86-ops.h

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1602]
After commit [5be2226f417d](https://git.kernel.org/next/linux-next/c/5be2226f417d5b06d17e6c52d6e341cf43c29e48) ("KVM: x86: allow defining return-0 static calls") in -next:

```
$ make -skj"$(nproc)" LLVM=1 mrproper allmodconfig arch/x86/kvm/x86.o
In file included from arch/x86/kvm/x86.c:19:
In file included from ./include/linux/kvm_host.h:45:
In file included from ./arch/x86/include/asm/kvm_host.h:1553:
./arch/x86/include/asm/kvm-x86-ops.h:82:1: error: pointer type mismatch ('bool (*)(struct kvm_vcpu *)' (aka '_Bool (*)(struct kvm_vcpu *)') and 'void *') [-Werror,-Wpointer-type-mismatch]
KVM_X86_OP_OPTIONAL_RET0(guest_apic_has_interrupt)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./arch/x86/include/asm/kvm_host.h:1551:54: note: expanded from macro 'KVM_X86_OP_OPTIONAL_RET0'
        static_call_update(kvm_x86_##func, kvm_x86_ops.func ? : \
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
./include/linux/static_call.h:154:42: note: expanded from macro 'static_call_update'
        typeof(&STATIC_CALL_TRAMP(name)) __F = (func);                  \
                                                ^~~~
In file included from arch/x86/kvm/x86.c:19:
In file included from ./include/linux/kvm_host.h:45:
In file included from ./arch/x86/include/asm/kvm_host.h:1553:
./arch/x86/include/asm/kvm-x86-ops.h:88:1: error: pointer type mismatch ('int (*)(struct kvm *, unsigned int)' and 'void *') [-Werror,-Wpointer-type-mismatch]
KVM_X86_OP_OPTIONAL_RET0(set_tss_addr)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./arch/x86/include/asm/kvm_host.h:1551:54: note: expanded from macro 'KVM_X86_OP_OPTIONAL_RET0'
        static_call_update(kvm_x86_##func, kvm_x86_ops.func ? : \
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
./include/linux/static_call.h:154:42: note: expanded from macro 'static_call_update'
        typeof(&STATIC_CALL_TRAMP(name)) __F = (func);                  \
                                                ^~~~
In file included from arch/x86/kvm/x86.c:19:
In file included from ./include/linux/kvm_host.h:45:
In file included from ./arch/x86/include/asm/kvm_host.h:1553:
./arch/x86/include/asm/kvm-x86-ops.h:89:1: error: pointer type mismatch ('int (*)(struct kvm *, u64)' (aka 'int (*)(struct kvm *, unsigned long long)') and 'void *') [-Werror,-Wpointer-type-mismatch]
KVM_X86_OP_OPTIONAL_RET0(set_identity_map_addr)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./arch/x86/include/asm/kvm_host.h:1551:54: note: expanded from macro 'KVM_X86_OP_OPTIONAL_RET0'
        static_call_update(kvm_x86_##func, kvm_x86_ops.func ? : \
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
./include/linux/static_call.h:154:42: note: expanded from macro 'static_call_update'
        typeof(&STATIC_CALL_TRAMP(name)) __F = (func);                  \
                                                ^~~~
In file included from arch/x86/kvm/x86.c:19:
In file included from ./include/linux/kvm_host.h:45:
In file included from ./arch/x86/include/asm/kvm_host.h:1553:
./arch/x86/include/asm/kvm-x86-ops.h:90:1: error: pointer type mismatch ('u64 (*)(struct kvm_vcpu *, gfn_t, bool)' (aka 'unsigned long long (*)(struct kvm_vcpu *, unsigned long long, _Bool)') and 'void *') [-Werror,-Wpointer-type-mismatch]
KVM_X86_OP_OPTIONAL_RET0(get_mt_mask)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./arch/x86/include/asm/kvm_host.h:1551:54: note: expanded from macro 'KVM_X86_OP_OPTIONAL_RET0'
        static_call_update(kvm_x86_##func, kvm_x86_ops.func ? : \
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
./include/linux/static_call.h:154:42: note: expanded from macro 'static_call_update'
        typeof(&STATIC_CALL_TRAMP(name)) __F = (func);                  \
                                                ^~~~
In file included from arch/x86/kvm/x86.c:19:
In file included from ./include/linux/kvm_host.h:45:
In file included from ./arch/x86/include/asm/kvm_host.h:1553:
./arch/x86/include/asm/kvm-x86-ops.h:108:1: error: pointer type mismatch ('bool (*)(struct kvm_vcpu *)' (aka '_Bool (*)(struct kvm_vcpu *)') and 'void *') [-Werror,-Wpointer-type-mismatch]
KVM_X86_OP_OPTIONAL_RET0(dy_apicv_has_pending_interrupt)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./arch/x86/include/asm/kvm_host.h:1551:54: note: expanded from macro 'KVM_X86_OP_OPTIONAL_RET0'
        static_call_update(kvm_x86_##func, kvm_x86_ops.func ? : \
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
./include/linux/static_call.h:154:42: note: expanded from macro 'static_call_update'
        typeof(&STATIC_CALL_TRAMP(name)) __F = (func);                  \
                                                ^~~~
5 errors generated.
```

Sean Christopherson sent https://lore.kernel.org/r/20220223162355.3174907-1-seanjc@google.com/.

I have a small reproducer in case this turns out to be a bug in clang: https://godbolt.org/z/P7nEdzejE

```c
struct static_call_key {
    void *func;
};

extern void __static_call_update(struct static_call_key *key, void *tramp, void *func);

struct kvm_vcpu;
struct kvm_x86_ops {
    _Bool (*guest_apic_has_interrupt)(struct kvm_vcpu *vcpu);
};
struct kvm_x86_ops kvm_x86_ops;

extern struct static_call_key __SCK__kvm_x86_guest_apic_has_interrupt;
extern typeof(*(((struct kvm_x86_ops *)0)->guest_apic_has_interrupt)) __SCT__kvm_x86_guest_apic_has_interrupt;

#define static_call_update(name, func) ({ \
    typeof(&__SCT__kvm_x86_##name) __F = (func); \
    __static_call_update(&__SCK__kvm_x86_##name, \
                         &__SCT__kvm_x86_##name, __F); \
});

long __static_call_return0(void)
{
    return 0;
}

void warn(void)
{
    static_call_update(guest_apic_has_interrupt, kvm_x86_ops.guest_apic_has_interrupt ? : (void *)__static_call_return0);
}

void no_warn(void) {
    if (kvm_x86_ops.guest_apic_has_interrupt) {
        static_call_update(guest_apic_has_interrupt, kvm_x86_ops.guest_apic_has_interrupt);
    } else {
        static_call_update(guest_apic_has_interrupt, (void *)__static_call_return0);
    }
}
```

```
$ clang -Wall -Wextra -fsyntax-only repro.c
repro.c:29:84: warning: pointer type mismatch ('_Bool (*)(struct kvm_vcpu *)' and 'void *') [-Wpointer-type-mismatch]
        static_call_update(guest_apic_has_interrupt, kvm_x86_ops.guest_apic_has_interrupt ? : (void *)__static_call_return0);
                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
repro.c:17:43: note: expanded from macro 'static_call_update'
    typeof(&__SCT__kvm_x86_##name) __F = (func); \
                                          ^~~~
1 warning generated.

$ gcc -Wall -Wextra -c -o /dev/null repro.c
```

**Comment by nathanchance** [comment]
Patch accepted: https://git.kernel.org/pub/scm/virt/kvm/kvm.git/commit/?id=925088781eede43cf6616a1197c31dee451b7948

**Comment by nathanchance** [comment]
Patch is now available in `next-20220307`: https://git.kernel.org/next/linux-next/c/925088781eede43cf6616a1197c31dee451b7948
